File: frontend/src/services/categoryService.ts
import api from "./api";
import type { CategoryResponse, CategoryRequest } from "../types/category";

export const categoryService = {
  findAll: async (): Promise<CategoryResponse[]> => {
    const response = await api.get<CategoryResponse[]>("/api/categories");
    return response.data;
  },

  findById: async (id: string): Promise<CategoryResponse> => {
    const response = await api.get<CategoryResponse>(`/api/categories/${id}`);
    return response.data;
  },

  create: async (data: CategoryRequest): Promise<CategoryResponse> => {
    const response = await api.post<CategoryResponse>("/api/categories", data);
    return response.data;
  },

  update: async (
    id: string,
    data: CategoryRequest,
  ): Promise<CategoryResponse> => {
    const response = await api.put<CategoryResponse>(
      `/api/categories/${id}`,
      data,
    );
    return response.data;
  },

  delete: async (id: string): Promise<void> => {
    await api.delete(`/api/categories/${id}`);
  },
};

File: backend/src/main/java/com/vitorsaucedo/finly/domain/category/CategoryService.java
package com.vitorsaucedo.finly.domain.category;

import com.vitorsaucedo.finly.domain.user.UserService;
import com.vitorsaucedo.finly.dto.request.CategoryRequest;
import com.vitorsaucedo.finly.dto.response.CategoryResponse;
import com.vitorsaucedo.finly.exception.BusinessException;
import com.vitorsaucedo.finly.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class CategoryService {

    private final CategoryRepository categoryRepository;
    private final UserService userService;

    @Transactional(readOnly = true)
    public List<CategoryResponse> findAll(UUID userId) {
        return categoryRepository.findAllByUserIdOrDefault(userId)
                .stream()
                .map(this::toResponse)
                .toList();
    }

    @Transactional(readOnly = true)
    public CategoryResponse findById(UUID id, UUID userId) {
        return categoryRepository.findByIdAndUserId(id, userId)
                .map(this::toResponse)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found"));
    }

    @Transactional
    public CategoryResponse create(CategoryRequest request, UUID userId) {
        Category category = Category.builder()
                .user(userService.getAuthenticatedUser(userId))
                .name(request.name())
                .type(request.type())
                .color(request.color())
                .icon(request.icon())
                .isDefault(false)
                .build();

        return toResponse(categoryRepository.save(category));
    }

    @Transactional
    public CategoryResponse update(UUID id, CategoryRequest request, UUID userId) {
        Category category = categoryRepository.findByIdAndUserId(id, userId)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found"));

        if (category.isDefault()) {
            throw new BusinessException("Default categories cannot be edited");
        }

        category.setName(request.name());
        category.setType(request.type());
        category.setColor(request.color());
        category.setIcon(request.icon());

        return toResponse(categoryRepository.save(category));
    }

    @Transactional
    public void delete(UUID id, UUID userId) {
        Category category = categoryRepository.findByIdAndUserId(id, userId)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found"));

        if (category.isDefault()) {
            throw new BusinessException("Default categories cannot be deleted");
        }

        categoryRepository.delete(category);
    }

    public Category getCategory(UUID id, UUID userId) {
        return categoryRepository.findByIdAndUserId(id, userId)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found"));
    }

    private CategoryResponse toResponse(Category category) {
        return new CategoryResponse(
                category.getId(),
                category.getName(),
                category.getType(),
                category.getColor(),
                category.getIcon(),
                category.isDefault(),
                category.getCreatedAt()
        );
    }
}